transport_nantes/photo/views.py

```diff
@@ -85,3 +96,266 @@ def get_context_data(self, **kwargs):
+
+
+class PhotoView(FormView):
+    """
+    View to display a single photo and up/down vote it
+    """
+    template_name = 'photo/single_entry.html'
+    form_class = AnonymousVoteForm
+
+    def get(self, request: HttpRequest, *args, **kwargs) -> HttpResponse:
+        """Overriding here so we can return HttpResponseForbidden if needed"""
+        try:
+            context_or_response = self.get_context_data()
+        except ForbiddenException as exception:
+            return HttpResponseForbidden(exception)
+        return self.render_to_response(context_or_response)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        photo_sha1 = self.kwargs.get("photo_sha1")
+        photo = get_object_or_404(PhotoEntry, sha1_name=photo_sha1)
+        tn_session_id = self.request.session.get("tn_session")
+        user = self.request.user if self.request.user.is_authenticated else None
+
+        if not photo.accepted:
+            # Author and authorized users can see the photo
+            if not any([
+                self.request.user == photo.user,
+                self.request.user.has_perm('photo.may_see_unaccepted_photos')
+            ]):
+                raise ForbiddenException(
+                    "Cette photo n'as pas encore été acceptée")
+        context["photo"] = photo
+
+        if photo.accepted:
+            self.set_next_and_previous_arrows(context, photo)
+            # Because it's possible that user is not authenticated (anonymous vote)
+            # we need to check if the user has already voted for this photo
+            # using its session id as well.
+            if Vote.objects.filter(
+                    Q(user=user)
+                    | Q(tn_session_id=tn_session_id),
+                    captcha_succeeded=True,
+            ).exists():
+                context["has_voted"] = True
+            else:
+                context["has_voted"] = False
+
+            # To set the proper styles on the vote buttons, we check if the
+            # user has already voted on this photo
+            last_vote = get_user_vote(
+                self.request.user, photo, tn_session_id=tn_session_id)
+
+            if last_vote:
+                if last_vote.vote_value is True:
+                    context["last_vote"] = "upvote"
+                elif last_vote.vote_value is False:
+                    context["last_vote"] = "downvote"
+
+            if not last_vote and user:
+                # If user is logged in but never voted, we ask for consent
+                # to send them emails
+                context["form"] = SimpleVoteFormWithConsent()
+
+        return context
+
+    def set_next_and_previous_arrows(self, context: dict, photo: 'PhotoEntry') -> None:
+        """
+        Set the next and previous photo arrows
+
+        context is passed by reference.
+        """
+        all_accepted_photos = list(PhotoEntry.objects.filter(accepted=True))
```

Comment:
This fetches all photos on every view in order to check which are previous and next.  Even if the user never clicks those buttons.

Can we not be lazy and have links for showing the next and the previous photo, with argument the current photo?  Then in those functions we do a query with the right order() and a first() or last().

There's a subtly about user feedback if it's already the first or the last photo.  You can get around that by making the list circular.

Reply:
Photo entries aren't very numerous so this can be a concern, I think this is a premature optimization. 
If we went for that, then an ajax call would probably work, but I don't think this is necessary at this point.

Reply:
```
def view_to_show_image(image_sha1):
    this_image = photos.get(sha1_name=image_sha1)
    if not this_image.accepted:
        # 404 here.
    try:
        next_image = photos.filter(
            accepted=True, sha1_name__gt=image_sha1).order_by("sha1_name").first()
    catch photo.DoesNotExist:
        # This must exist, barring race conditions, because this_image is accepted.
        next_image = photos.filter(accepted=True).order_by("sha1_name").first()
    try:
        prev_image = photos.filter(
            accepted=True, sha1_name__lt=image_sha1).order_by("-sha1_name").first()
    catch photo.DoesNotExist:
        # This must exist, barring race conditions, because this_image is accepted.
        next_image = photos.filter(accepted=True).order_by("-sha1_name").first()

```

Reply:
Do you want me to refactor ?

Reply:
Yes, that was kinda the point...  ;-)

Whenever you have to disable an element that you display or usually display, it's good user interface design to ask yourself if a slight change in metaphor could make the flow easier.